webui/react/src/hooks/useSettings.ts

```diff
@@ -0,0 +1,167 @@
+
+const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
+  if (param === null) return undefined;
+  if (type === BaseType.Boolean) return param === 'true';
+  if (type === BaseType.Float || type === BaseType.Integer) {
+    const value = type === BaseType.Float ? parseFloat(param) : parseInt(param);
+    return !isNaN(value) ? value : undefined;
+  }
+  if (type === BaseType.String) return param;
+  return undefined;
+};
+
+const queryToSettings = <T>(
+  query: string,
+  config: SettingsConfig,
+  defaultSettings?: GenericSettings,
+): T => {
+  const params = queryString.parse(query);
+  return config.settings.reduce((acc, prop) => {
+    /*
+     * Attempt to decode the query parameter and if anything
+     * goes wrong, set it to the default value.
+     */
+    try {
+      const paramValue = params[prop.key];
+      const baseType = prop.type.baseType;
+
+      /*
+       * Convert the string-based query params to primitives.
+       * `undefined` values can happen if the query param values are invalid.
+       *   string[] => Primitive[] | undefined
+       *   string   => Primitive | undefined
+       *   null     => undefined
+       */
+      const queryValue = Array.isArray(paramValue)
+        ? paramValue
+          .map(value => queryParamToType(baseType, value))
+          .filter((value): value is Primitive => value !== undefined)
+        : queryParamToType(baseType, paramValue);
+
+      /*
+       * When expecting an array, convert valid non-array values into an array.
+       * Example - 'PULLING' => [ 'PULLING' ]
+       */
+      const normalizedValue = prop.type.isArray && queryValue != null && !Array.isArray(queryValue)
+        ? [ queryValue ] : queryValue;
+
+      acc[prop.key] = normalizedValue;
+    } catch (e) {
+      if (defaultSettings && defaultSettings[prop.key] !== undefined) {
+        acc[prop.key] = defaultSettings[prop.key];
+      }
+    }
+    return acc;
+  }, {} as GenericSettings) as unknown as T;
+};
+
+const useSettings = <T>(
+  config: SettingsConfig,
+  basePath: string,
+): SettingsHook<T> => {
+  const history = useHistory();
+  const location = useLocation();
+  const storage = useStorage(config.storagePath);
+  const prevSearch = usePrevious(location.search, undefined);
+  const [ settings, setSettings ] = useState<T>(() => {
+    return config.settings.reduce((acc, config) => {
+      let value = config.defaultValue;
+
+      // Pull from storage if config storage key is defined.
+      if (config.storageKey) {
+        value = storage.getWithDefault(config.storageKey, value);
+      }
+      acc[config.key] = value;
+      return acc;
+    }, {} as GenericSettings) as unknown as T;
+  });
+
+  const filterCount = useMemo(() => {
+    return config.settings.reduce((acc, config) => {
+      const settingsValue = (settings as unknown as GenericSettings)[config.key];
+      return acc + (config.isFilter && settingsValue !== config.defaultValue ? 1 : 0);
```

Comment:
we might need to check the values here as this comparison would only compare by reference for non premitives and fail for say `[1] === [1]`

Reply:
great catch! this should be able to handle more than just primitives.